### PR TITLE
time-based filtering

### PR DESCRIPTION
**Pull request by robcza**
filter expert modification and test, as discussed previously in here: https://github.com/certtools/intelmq/issues/320

Parameters for time based filtering:
* not_before - events before this time will be dropped
* not_after - events after this time will be dropped

Both parameters accept string values describing absolute or relative time:
* absolute
 * basically anything parsable by datetime parser, eg. "2015-09-012T06:22:11+00:00"
 * time.source taken from the event will be compared to this value to decide the filter behavior
* relative
 * accepted string formatted like this "<integer> <epoch>", where epoch could be any of following strings (could optionally end with trailing 's'): hour, day, week, month, year
 * time.source taken from the event will be compared to the value (now - relative) to decide the filter behavior

Examples of time filter definition:
* ```"not_before" : "2015-09-012T06:22:11+00:00"``` events older than the specified time will be dropped
* ```"not_after" : "6 months"``` just events older than 6 months will be passed through the pipeline

### CODE REVIEW

**Comment by SYNchroACK** [comment]
@robcza Awesome! Thank you very much.
Please check this link https://travis-ci.org/certtools/intelmq/jobs/83119441 to fix the pull req.

**Comment by robcza** [comment]
will do :)

File: intelmq/bots/experts/filter/expert.py [hunk 1]
```diff
@@ -3,20 +3,75 @@
 import sys
 
 from intelmq.lib.bot import Bot
+from intelmq.lib.harmonization import DateTime
+from datetime import datetime, timedelta
+from dateutil import parser
+import re
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: Is a built-in packages, should go in line before sys-import. You may use `isort -rc filename` (python package from pip)

File: intelmq/bots/experts/filter/expert.py [hunk 2]
```diff
@@ -3,20 +3,75 @@
...
 
+class FilterExpertBot(Bot):
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: PEP8: 2 newlines before definitions

File: intelmq/tests/bots/experts/filter/test_expert.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+import json
+import unittest
+
+import intelmq.lib.test as test
+from intelmq.bots.experts.filter.expert import FilterExpertBot
+
```
Comment (sebix) [inline intelmq/tests/bots/experts/filter/test_expert.py]: second newline missing here.

**Comment by robcza** [comment]
@SYNchroACK passed! :)
@sebix thank you for the comments, using pyCharm now to help me with the pep8 issues

**Comment by sebix** [comment]
With #344 the bots docs have been migrated from separate READMEs to the central docs/BOTS.md file.

I will test the code on Tuesday.

**Comment by mauroasilva** [comment]
Sorry for the late reply, but I think the time key used for filtering should be configurable. Something like time_filter_key, which could be any of the timestamps present on the message. It makes the filtering more flexible and it's pretty easy to replace the current static field for a parameter.

**Comment by aaronkaplan** [comment]
how's the status on this @robcza ? Do you think you are ready for the PR? Can we test?

File: intelmq/bots/experts/filter/expert.py [hunk 4]
```diff
@@ -1,22 +1,79 @@
...
+from dateutil import parser
 from intelmq.lib.bot import Bot
+from intelmq.lib.harmonization import DateTime
+
 
+class FilterExpertBot(Bot):
 
-class FilterBot(Bot):
+    # number of minutes in time units
+    timespans = ({'hour': 60, 'day': 24*60, 'week': 7*24*60, 'month': 30*24*60, 'year': 365*24*60})
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: Split into multiple lines and remove the redundant round brackets.

File: intelmq/bots/experts/filter/expert.py [hunk 5]
```diff
@@ -1,22 +1,79 @@
...
+        if len(result) == 1 and len(result[0]) == 2 and result[0][1] in FilterExpertBot.timespans:
+            return int(result[0][0]) * FilterExpertBot.timespans[result[0][1]]
+        else:
+            raise AttributeError("Could not process result of regex for attribute " + repr(relative_time))
+
+    # decide format of timefilter value and parse it
+    def parse_timeattr(self, time_attr):
+        try:
+            absolute = parser.parse(time_attr)
+        except:
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: Only catch ValueError here.

File: intelmq/bots/experts/filter/expert.py [hunk 6]
```diff
@@ -1,22 +1,79 @@
...
+    # number of minutes in time units
+    timespans = ({'hour': 60, 'day': 24*60, 'week': 7*24*60, 'month': 30*24*60, 'year': 365*24*60})
+
+    # parse relative time attributes
+    @staticmethod
+    def parse_relative(relative_time):
+        try:
+            result = re.findall(r'(\d+)\s+(\w+)?s', relative_time, re.UNICODE)
+        except Exception as e:
+            raise AttributeError("Could not apply regex to attribute \"%s\" with exception %s",
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: This should be ValueError, also on line 29 and 38.

File: intelmq/bots/experts/filter/expert.py [hunk 7]
```diff
@@ -1,22 +1,79 @@
...
+    def parse_timeattr(self, time_attr):
+        try:
+            absolute = parser.parse(time_attr)
+        except:
+            try:
+                relative = timedelta(minutes=self.parse_relative(time_attr))
+            except Exception as e:
+                self.logger.error("Could not parse \"%s\" as neither absolute or relative time, "
+                                  "exception raised: %s", time_attr, repr(e.args))
+                self.stop()
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: Error handling is done by the bots class, just raise the exception here.

File: intelmq/bots/experts/filter/expert.py [hunk 8]
```diff
@@ -26,26 +83,57 @@ def process(self):
             self.acknowledge_message()
             return
 
-        if self.parameters.filter_action == "drop":
+        # time based filtering
+        if event.contains('time.source'):
+            try:
+                event_time = parser.parse(event.value('time.source')).replace(tzinfo=pytz.timezone('UTC'))
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: pytz is never imported

File: intelmq/tests/bots/experts/filter/test_expert.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+import json
```
Comment (sebix) [inline intelmq/tests/bots/experts/filter/test_expert.py]: json is not used here

File: intelmq/tests/bots/experts/filter/test_expert.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+class TestFilterExpertBot(test.BotTestCase, unittest.TestCase):
+    """
+    A TestCase for FilterExpertBot.
+    """
+
+    @classmethod
+    def set_bot(self):
+        self.bot_reference = FilterExpertBot
+        self.sysconfig = {'not_before': '2014-01-01T00:00:00+00:00',
```
Comment (sebix) [inline intelmq/tests/bots/experts/filter/test_expert.py]: trailing whitespace

**Comment by sebix** [comment]
The tests are not complete. You only test one possibility. There are relative and absolute times and the result can also filter or keep. Results in 4 tests minimum. Then there's before and after, should both be at least tested once.

**Comment by robcza** [comment]
Thank you @sebix I am aware of the incomplete tests, however, I had problems initializing the bot for the test for four different configurations. Could you please advice how to achieve that?

I'm going to fix the inline comments as well

**Comment by sebix** [comment]
You can set `self.sysconfig = {'foo': 'bar'}` before running the bot using `self.run_bot()` in each test.

**Comment by robcza** [comment]
Bah! :( locally the tests go through n1. Could there be an issue with the same class defined in multiple test scripts? Seemed to me as a nice way to describe different cases.

**Comment by SYNchroACK** [comment]
@robcza Im not a unittest expert but It seems to me that you forge to add 'EXAMPLE_OUTPUT' in multiple files.

**Comment by robcza** [comment]
The problem does not seem to be in the input output events, but in the bot sysconfig:
```
test-bot: DEBUG: System configuration: parameter 'not_before' loaded  with value '2014-01-01T00:00:00+00:00'.
test-bot: DEBUG: System configuration: parameter 'not_after' loaded  with value '1 month'.
```

This is the same bot configured with two different values, however none of the tests assigns two different attributes to a test. @sebix can I ask you for a help. I'm worried the way I store individual tests is not right.

File: intelmq/bots/experts/filter/expert.py [hunk 11]
```diff
@@ -1,22 +1,78 @@
...
 from __future__ import unicode_literals
+
+import re
 import sys
+import pytz
+from datetime import datetime, timedelta
 
+from dateutil import parser
 from intelmq.lib.bot import Bot
+from intelmq.lib.harmonization import DateTime
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: imported, but unused

File: intelmq/bots/experts/filter/expert.py [hunk 12]
```diff
@@ -1,22 +1,78 @@
...
+
+    # decide format of timefilter value and parse it
+    def parse_timeattr(self, time_attr):
+        try:
+            absolute = parser.parse(time_attr)
+        except:
+            try:
+                relative = timedelta(minutes=self.parse_relative(time_attr))
+            except Exception as e:
+                raise e
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: Why do you catch the error and raise it immediatly?

File: intelmq/bots/experts/filter/expert.py [hunk 13]
```diff
@@ -1,22 +1,78 @@
...
 
-class FilterBot(Bot):
+    # number of minutes in time units
+    timespans = ({'hour': 60, 'day': 24*60, 'week': 7*24*60, 'month': 30*24*60, 'year': 365*24*60})
+
+    # parse relative time attributes
+    @staticmethod
+    def parse_relative(relative_time):
+        try:
+            result = re.findall(r'(\d+)\s+(\w+[^s])s?', relative_time, re.UNICODE)
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: `'-2 months'` matches and becomes `[('2', 'month')]`. Add markers for begin and end of string: `r'^(\d+)\s+(\w+[^s])s?$'`.

File: intelmq/bots/experts/filter/expert.py [hunk 14]
```diff
@@ -1,22 +1,78 @@
...
+        except Exception as e:
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: Which exceptions do you expect here?

File: intelmq/bots/experts/filter/expert.py [hunk 15]
```diff
@@ -1,22 +1,78 @@
...
+            raise AttributeError("Could not apply regex to attribute \"%s\" with exception %s",
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: Definitely not an Attribute error. If the input is invalid, it's a ValueError. See also comment above, which error can occur here?

File: intelmq/bots/experts/filter/expert.py [hunk 16]
```diff
@@ -1,22 +1,78 @@
...
+                                 repr(relative_time), repr(e.args))
+        if len(result) == 1 and len(result[0]) == 2 and result[0][1] in FilterExpertBot.timespans:
+            return int(result[0][0]) * FilterExpertBot.timespans[result[0][1]]
+        else:
+            raise AttributeError("Could not process result of regex for attribute " + repr(relative_time))
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: As above, this is a ValueError (invalid value in input):
https://docs.python.org/3/library/exceptions.html#AttributeError
> exception AttributeError

> Raised when an attribute reference (see Attribute references) or assignment fails. (When an object does not support attribute references or attribute assignments at all, TypeError is raised.)

File: intelmq/tests/bots/experts/filter/test_expert_relative_after.py [hunk 17]
```diff
@@ -0,0 +1,34 @@
...
+
+    @classmethod
+    def set_bot(self):
+        self.bot_reference = FilterExpertBot
+        self.input_message = EXAMPLE_INPUT
+        self.sysconfig = {'not_after': '1 month'}
+
+    def test_Relative_After(self):
+        self.run_bot()
+        self.assertOutputQueueLen(1)
```
Comment (sebix) [inline intelmq/tests/bots/experts/filter/test_expert_relative_after.py]: Please use `self.assertMessageEqual(0, EXAMPLE_INPUT)`, so also the content of the message is checked. (Input = output here)

**Comment by robcza** [comment]
@sebix Thank you for the comments, appreciate it

File: intelmq/bots/experts/filter/expert.py [hunk 18]
```diff
@@ -1,22 +1,74 @@
...
+    timespans = {'hour': 60, 'day': 24*60, 'week': 7*24*60,
+                 'month': 30*24*60, 'year': 365*24*60}
+
+    # parse relative time attributes
+    @staticmethod
+    def parse_relative(relative_time):
+        try:
+            result = re.findall(r'^(\d+)\s+(\w+[^s])s?$', relative_time, re.UNICODE)
+        except Exception as e:
+            raise AttributeError("Could not apply regex to attribute \"%s\" with exception %s",
```
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: ValueError please here too.

**Comment by robcza** [comment]
Hallelujah, thank you for the help with test @sebix 
Is it ok, to keep the adjusted test.py lib in this pull request?

File: intelmq/bots/experts/filter/expert.py (hunk 5 above)
Comment (sebix) [inline intelmq/bots/experts/filter/expert.py]: As already said, we don't want to catch all Exceptions here (no IOError, MemoryError, OverflowError, SystemError, SystemExit etc.)

**Comment by sebix** [comment]
The code itself is fine, but many lines are still too long. There's also still one too broad except-clause.

**Comment by robcza** [comment]
@sebix Ok, I understand the exception handling process now and fixed the broad except-clauses.

One thing I don't really understand is the line length culprit. My pyCharm studio says line length is pep8 compliant, you say it is not - who is right?

**Comment by aaronkaplan** [comment]
> On 09 Dec 2015, at 13:19, Robert Šefr <notifications@github.com> wrote:
> 
> @sebix Ok, I understand the exception handling process now and fixed the broad except-clauses.
> 
> One thing I don't really understand is the line length culprit. My pyCharm studio says line length is pep8 compliant, you say it is not - who is right?
> 
Does your pyCharm read a .pep8 file which I added?

I have to admit, I don't care so much if its' 80 chars / line or not. for me this is a good exception to full PEP8 compliancy.
Some of the code will look extremely ugly if you *have to* break a line at 80 chars. See for example the n6stomp parser.

**Comment by robcza** [comment]
I have 120 columns right now and seems good for readability on github (github has 126 I believe), but is not fully compliant with pep8 as stated in developers guide in documentation, however there is this .pep8 ```max-line-length = 132```

I vote for 120, however I'll be happy with any other reasonable final statement to have this sorted out.

**Comment by aaronkaplan** [comment]
> On 09 Dec 2015, at 14:51, Robert Šefr <notifications@github.com> wrote:
> 
> I have 120 columns right now and seems good for readability on github (github has 126 I believe), but is not fully compliant with pep8 as stated in developers guide in documentation, however there is this .pep8 max-line-length = 132
> 
> I vote for 120, however I'll be happy with any other reasonable final statement to have this sorted out.
> 

good enough for me.

**Pull request merged**